### PR TITLE
Make all features explicit

### LINKED ISSUE #5368: remove hidden feature flags
In client we have a feature flags, that _aren't_ exposed through their respective `Cargo.toml`s, like in sc-client-db, we have the 'hidden' `kvdb-rocksdb`-feature that some (like sc-service) activate but isn't known otherwise. The reason for that seems to be that the name of the feature is the same as for a dependency, which cargo prevents you from having. 

But not exposing the feature means that ~~neither `crates.io` nor `docs.rs` know about it and~~ they are harder to discover and thus go more often unnoticed. This is not the common convention. Instead we should rename the features and make them explicit.

refs #5367

### PR DESCRIPTION
**Pull request by pscott**
This PR makes "hidden" features (i.e. features that were used in the code for cfg but that were not exposed in the corresponding `Cargo.toml` files)  more explicit.

When features have the same name as dependencies, they are suffixed with `with-`.

Closes #5368

### CODE REVIEW

File: client/transaction-pool/Cargo.toml [hunk 1]
```diff
@@ -38,3 +38,7 @@ hex = "0.4"
 sp-keyring = { version = "2.0.0-rc2", path = "../../primitives/keyring" }
 substrate-test-runtime-transaction-pool = { version = "2.0.0-rc2", path = "../../test-utils/runtime/transaction-pool" }
 substrate-test-runtime-client = { version = "2.0.0-rc2", path = "../../test-utils/runtime/client" }
+
+
+[features]
+test-helpers = []
```
Comment (bkchr) [inline client/transaction-pool/Cargo.toml:44]: Please don't add another one of this

File: primitives/runtime-interface/proc-macro/Cargo.toml [hunk 2]
```diff
@@ -21,3 +21,7 @@ quote = "1.0.3"
 proc-macro2 = "1.0.3"
 Inflector = "0.11.4"
 proc-macro-crate = "0.1.4"
+
+[features]
+default = ["std"]
```
Comment (bkchr) [inline primitives/runtime-interface/proc-macro/Cargo.toml:26]: The macro does not need this

File: frame/support/procedural/Cargo.toml [hunk 3]
```diff
@@ -19,3 +19,7 @@ frame-support-procedural-tools = { version = "2.0.0-rc2", path = "./tools" }
 proc-macro2 = "1.0.6"
 quote = "1.0.3"
 syn = { version = "1.0.7", features = ["full"] }
+
+[features]
+default = ["std"]
```
Comment (bkchr) [inline frame/support/procedural/Cargo.toml:24]: Same

File: bin/node/cli/Cargo.toml [hunk 4]
```diff
@@ -156,3 +156,4 @@ runtime-benchmarks = [
 	"node-runtime/runtime-benchmarks",
 	"frame-benchmarking-cli",
 ]
+rhd = ["default"]
```
Comment (bkchr) [inline bin/node/cli/Cargo.toml:159]: We don't have rhd anywhere in the codebase

Comment (pscott) [reply bin/node/cli/Cargo.toml:159]: Yup for some reason I clearly remember having it pop out somewhere... gonna remove this one and keep investigating :)

**Comment by pscott** [comment]
@bkchr I added those because those features were used inside our code base but not exposed in the Cargo.toml. For example for the `test-helpers`
https://github.com/paritytech/substrate/blob/606b4faf2f52ec588235d3cdff6c7364a393a645/client/transaction-pool/src/lib.rs#L31

and for `std`: 
https://github.com/paritytech/substrate/blob/606b4faf2f52ec588235d3cdff6c7364a393a645/frame/support/procedural/src/storage/metadata.rs#L97

Should we not declare those as features and do something else instead? Or leave it as it?

File: bin/node/cli/Cargo.toml (hunk 4 above)
Comment (pscott) [reply bin/node/cli/Cargo.toml]: Oh I know why! We previously had a `#[cfg(feature = "rhd")]` here: https://github.com/paritytech/substrate/blob/606b4faf2f52ec588235d3cdff6c7364a393a645/bin/node/cli/src/service.rs#L436 but it has now been removed :)

**Comment by pscott** [comment]
Btw I used this [little program I wrote](https://github.com/pscott/cargo-features) to help me find the hidden features :)

**Comment by bkchr** [comment]
> @bkchr I added those because those features were used inside our code base but not exposed in the Cargo.toml. For example for the `test-helpers`
> 
> https://github.com/paritytech/substrate/blob/606b4faf2f52ec588235d3cdff6c7364a393a645/client/transaction-pool/src/lib.rs#L31
>

Remove test-helpers there.


> and for `std`:
> 
> https://github.com/paritytech/substrate/blob/606b4faf2f52ec588235d3cdff6c7364a393a645/frame/support/procedural/src/storage/metadata.rs#L97
> 

This is generated code.

File: client/db/src/utils.rs [hunk 5]
```diff
@@ -255,7 +255,7 @@ pub fn open_database<Block: BlockT>(
 				.map_err(|err| sp_blockchain::Error::Backend(format!("{}", err)))?;
 			sp_database::as_database(db)
 		},
-		#[cfg(not(any(feature = "kvdb-rocksdb", test)))]
+		#[cfg(not(any(feature = "with-kvdb-rocksdb", test)))]
 		DatabaseSettingsSrc::RocksDb { .. } => {
 			return db_open_error("kvdb-rocksdb");
```
Comment (bkchr) [inline client/db/src/utils.rs:260]:
```suggestion
			return db_open_error("with-kvdb-rocksdb");
```

File: client/db/Cargo.toml [hunk 6]
```diff
@@ -47,3 +47,6 @@ tempfile = "3"
 [features]
 default = []
 test-helpers = []
+with-kvdb-rocksdb = ["kvdb-rocksdb"]
+with-parity-db = ["parity-db"]
+subdb = []
```
Comment (bkchr) [inline client/db/Cargo.toml:52]:
```suggestion
with-subdb = []
```

File: client/db/src/utils.rs [hunk 7]
```diff
@@ -268,12 +268,12 @@ pub fn open_database<Block: BlockT>(
 		DatabaseSettingsSrc::SubDb { .. } => {
 			return db_open_error("subdb");
```
Comment (bkchr) [inline client/db/src/utils.rs:269]:
```suggestion
			return db_open_error("with-subdb");
```

File: client/db/src/utils.rs [hunk 8]
```diff
@@ -268,12 +268,12 @@ pub fn open_database<Block: BlockT>(
...
-		#[cfg(feature = "parity-db")]
+		#[cfg(feature = "with-parity-db")]
 		DatabaseSettingsSrc::ParityDb { path } => {
 			crate::parity_db::open(&path)
 				.map_err(|e| sp_blockchain::Error::Backend(format!("{:?}", e)))?
 		},
-		#[cfg(not(feature = "parity-db"))]
+		#[cfg(not(feature = "with-parity-db"))]
 		DatabaseSettingsSrc::ParityDb { .. } => {
 			return db_open_error("parity-db");
```
Comment (bkchr) [inline client/db/src/utils.rs:278]:
```suggestion
			return db_open_error("with-parity-db");
```

File: client/executor/runtime-test/Cargo.toml [hunk 9]
```diff
@@ -31,3 +31,4 @@ std = [
 	"sp-std/std",
 	"sp-allocator/std",
 ]
+strict = ["default"]
```
Comment (bkchr) [inline client/executor/runtime-test/Cargo.toml:34]: Remove this and this weird "cfg_attr" in the lib.rs

**Comment by pscott** [comment]
I see that [the polkadot check](https://gitlab.parity.io/parity/substrate/builds/540306) fails, however I don't think this PR needs a companion PR, so I don't think there's anything I can do here :)

File: client/executor/runtime-test/src/lib.rs [hunk 10]
```diff
@@ -1,5 +1,4 @@
 #![cfg_attr(not(feature = "std"), no_std)]
-#![cfg_attr(feature = "strict", deny(warnings))]
```
Comment (expenses) [inline client/executor/runtime-test/src/lib.rs:2]: Why did you remove this?

File: client/db/Cargo.toml [hunk 11]
```diff
@@ -47,3 +47,6 @@ tempfile = "3"
 [features]
 default = []
 test-helpers = []
+with-kvdb-rocksdb = ["kvdb-rocksdb"]
+with-parity-db = ["parity-db"]
+with-subdb = []
```
Comment (expenses) [inline client/db/Cargo.toml:52]: I like this `with-` naming scheme :^)

File: client/transaction-pool/src/lib.rs [hunk 12]
```diff
@@ -28,7 +28,7 @@ mod metrics;
 
 pub mod error;
 
-#[cfg(any(feature = "test-helpers", test))]
+#[cfg(test)]
```
Comment (expenses) [inline client/transaction-pool/src/lib.rs:31]: Changing this means that we can't use the stuff in `testing` outside of this crate. Not sure if it is used though.

File: client/executor/runtime-test/src/lib.rs (hunk 10 above)
Comment (pscott) [reply client/executor/runtime-test/src/lib.rs:2]: [See this comment](https://github.com/paritytech/substrate/pull/6267#discussion_r436165804)

File: client/db/Cargo.toml (hunk 11 above)
Comment (pscott) [reply client/db/Cargo.toml:52]: Props to @gnunicorn for the naming convention :p

File: client/transaction-pool/src/lib.rs (hunk 12 above)
Comment (pscott) [reply client/transaction-pool/src/lib.rs:31]: [See this comment](https://github.com/paritytech/substrate/pull/6267#issuecomment-639541465) :D
